File: archive/d/dart/sleep-sort.dart
import 'dart:async';

void main(List<String> args) async {
  
  if (args.length == 0 || args[0].isEmpty || args[0].split(",").length == 1) {
    print('Usage: please provide a list of at least two integers to sort in the format "1, 2, 3, 4, 5"');
  } else {
    List<int> userInput = args[0].split(",").map((str) => int.tryParse(str))
      .takeWhile((test) => test != null)
      .toList();

    List<int> sorted = await sleepsort(userInput);

    print(sorted);
  }
}

Future<List<int>> sleepsort(Iterable<int> input) async {
  List<int> sorted = List();
  await Future.wait(input.map((i) => Future.delayed(Duration(seconds: i), () => sorted.add(i))));
  return Future.value(sorted);
}

File: archive/s/scala/SleepSort.scala
import java.util.concurrent.ConcurrentLinkedQueue
import scala.concurrent.ExecutionContext.Implicits.global
import scala.concurrent.duration.*
import scala.concurrent.{Await, Future}
import scala.jdk.CollectionConverters.*
import scala.util.Try

object SleepSort:
  private val TimeUnit = 100.milliseconds
  private val Buffer = 500.milliseconds

  private val usage =
    """Usage: please provide a list of at least two integers in the format "1, 2, 3, 4, 5""""

  def main(args: Array[String]): Unit =
    val result =
      args.headOption
        .flatMap(parse)
        .filter(_.length >= 2)
        .map(sort)
        .map(_.mkString(", "))
        .getOrElse(usage)

    println(result)

  private def parse(input: String): Option[List[Int]] =
    Try(input.split(',').map(_.trim.toInt.max(0)).toList).toOption

  private def sort(xs: List[Int]): List[Int] =
    val collector = new ConcurrentLinkedQueue[Int]()

    val tasks = xs.map { n =>
      Future {
        Thread.sleep((n * TimeUnit).toMillis)
        collector.add(n)
      }
    }

    // The buffer is there to ensure that all tasks have finished
    val maxWait = (xs.maxOption.getOrElse(0) * TimeUnit) + Buffer
    val allDone = Future.sequence(tasks)
    Await.result(allDone, maxWait)

    collector.asScala.toList